File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
'''
# 121. Best Time to Buy and Sell Stock

use **bottom-up dynamic programming** to solve the problem.

## Time and Space Complexity
```
TC: O(n)
SC: O(1)
```

## TC is O(n):
- iterating through the list just once to find the maximum profit. = O(n)

## SC is O(1):
- using two variables to store the minimum price and maximum profit. = O(1)
'''

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # is there's a base case??
        if len(prices) == 1:
            return 0
        
        # two pointers
        lowest_price = prices[0]
        max_profit = 0
        
        for price in prices:
            lowest_price = min(price, lowest_price)
            curr_profit = price - lowest_price
            max_profit = max(curr_profit, max_profit)
        
        return max_profit

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int min_price = prices[0];
        int max_profit = 0;

        for (int price : prices){
            min_price = Math.min(price, min_price);
            max_profit = Math.max(price - min_price, max_profit);
        }

        return max_profit;
    }
}